src/parser/ts/grammar.ne

```diff
@@ -0,0 +1,196 @@
+}
+
+const extract2Elements = d => {
+	return [d[0], d[2]]
+}
+
+const extractOperator = ([s, operator, S]) => {
+  return { operator }
+}
+
+const extractNumber = ([constant]) => {
+  return { constant }
+}
+
+const extractPI = d => {
+  return { constant: d[0] + d[2] }
+}
+
+const extractFunction = ([func, br, args, BR]) => {
+  if (args && args.length) {
+    return { func: func || '', args }
+  }
+  return { func }
+}
+
+const extractVariable = ([_, d]) => "$" + d;
+
+const emptyObject = () => ({});
+const emptyArray = () => ([]);
+%}
+
+query -> _ trimmed _ {% d => d[1] %}
+
+trimmed -> compositeElement {% id %}
+  | function {% id %}
+
+function -> unary br arithmethicElements BR {% extractFunction %}
+  | unary br oneElement BR {% extractFunction %}
+  | binary br twoElements BR {% extractFunction %}
+  | n_ary br commaSeparatedElements BR {% extractFunction %}
+  | "map" br map_func_args BR {% extractFunction %}
+
+oneElement -> element {% extractCommaSeparatedArray %}
+twoElements -> compositeElement comma compositeElement {% extract2Elements %}
+commaSeparatedElements -> compositeElement (comma compositeElement):* {% extractCommaSeparatedArray %}
+arithmethicElements -> element (operator element):+ {% extractOperationsArray %}
+
+map_func_args -> compositeElement "," array "," array comma number {% extractMapFuncArgs %}
+
+compositeElement -> element {% id %}
+  | arithmethicElements {% id %}
+
+element -> function {% id %}
+  | type condition {% formatQuery %}
+  | number {% extractNumber %}
+  | "pi(" _ ")" {% extractPI %}
+
+type -> "ts"i {% ([d]) => d.toLowerCase() %}
+
+condition -> curl CURL {% emptyArray %}
+  | curl pair (comma pair):* CURL {% extractConditionToArray %}
+
+filter -> "!=" {% id %}
+  | "=~" {% id %}
+  | "!~" {% id %}
+equals -> "=" {% id %}
+prop_name -> [A-Za-z0-9_]:+ {% join %}
+
+string -> sqstring {% id %}
+  | dqstring {% id %}
+number -> decimal {% id %}
+array -> sqr SQR {% emptyArray %}
+  | sqr value (comma value):* SQR {% extractArray %}
+object -> curl CURL {% emptyObject %}
+  | curl pair (comma pair):* CURL {% extractObject %}
+pair -> prop_name equals value
+  | prop_name filter string
+value ->
+  object {% id %}
+  | array {% id %}
+  | number {% id %}
+  | string {% id %}
+  | variable {% id %}
+  | "true" {% () => true %}
+  | "false" {% () => false %}
+  | "null" {% () => null %}
+
+variable -> "$" prop_name {% extractVariable %}
+  | "[[" prop_name "]]" {% extractVariable %}
+
+operator -> _ "+" _ {% extractOperator %}
+  | _ "-" _ {% extractOperator %}
+  | _ "/" _ {% extractOperator %}
+  | _ "*" _ {% extractOperator %}
+comma -> _ "," _ {% d => d[1] %}
+unary -> "sin" {% id %}
+  | "cos" {% id %}
+  | "ln" {% id %}
+  | "sqrt" {% id %}
+  | "exp" {% id %}
+  | "abs" {% id %}
+  | null {% id %}
+binary -> "pow" {% id %}
+  | "round" {% id %}
+  | "on_error" {% id %}
+n_ary -> "max" {% id %}
+  | "min" {% id %}
+  | "avg" {% id %}
+  | "sum" {% id %}
+br ->  _ "(" {% null %}
```

Comment:
looks a bit weird to me, harder to read

Reply:
the reason I have is that I don't need to have a bunch of space parameters in extractor (which is even worse to read).

Maybe I could rename it or something?

Reply:
I see your point. No, don't think you'll find a better name, let's leave it as is